toponetx/utils/normalization.py

```diff
@@ -0,0 +1,359 @@
+"""Normalize of adjacency, boundary matrices of complexes."""
+
+import numpy as np
+import scipy.sparse.linalg as spl
+from numpy import ndarray
+from numpy.linalg import pinv
+from scipy.sparse import coo_matrix, diags, identity
+
+__all__ = [
+    "normalize_laplacian",
+    "normalize_x_laplacian",
+    "kipf_adjacency_matrix_normalization",
+    "asymmetric_xu_normalization",
+    "bunch_normalization",
+    "_compute_B1_normalized",
+    "_compute_B1T_normalized",
+    "_compute_B2_normalized",
+    "_compute_B2T_normalized",
+    "_compute_D1",
+    "_compute_D2",
+    "_compute_D3",
+    "_compute_D5",
+]
+
+
+def normalize_laplacian(L):
+    """Normalize the Laplacian matrix.
+
+    Parameters
+    ----------
+        L (scipy.sparse.csr_matrix): The Laplacian matrix.
+
+    Returns
+    -------
+        scipy.sparse.csr_matrix: The normalized Laplacian matrix.
+
+    Notes
+    -----
+        This function normalizes the Laplacian matrix by dividing it by the largest eigenvalue.
+    """
+    topeigen_val = spl.eigsh(L.asfptype(), k=1, which="LM", return_eigenvectors=False)[
+        0
+    ]
+    out = L.copy()
+    out *= 1.0 / topeigen_val
+    return out
+
+
+def normalize_x_laplacian(L, Lx):
+    """Normalize the up or down Laplacians.
+
+    Parameters
+    ----------
+        L (scipy.sparse.csr_matrix): The Laplacian matrix.
+        Lx (scipy.sparse.csr_matrix): The up or down Laplacian matrix.
+
+    Returns
+    -------
+        scipy.sparse.csr_matrix: The normalized up or down Laplacian matrix.
+
+    Notes
+    -----
+        This function normalizes the up or down Laplacian matrices by dividing it
+        by the largest eigenvalue of the Laplacian matrix.
+    """
+    assert L.shape[0] == L.shape[1]
+    topeig = spl.eigsh(L.asfptype(), k=1, which="LM", return_eigenvectors=False)[0]
+    out = Lx.copy()
+    out *= 1.0 / topeig
+    return out
+
+
+def kipf_adjacency_matrix_normalization(
```

Comment:
I think we should use descriptive names for these functions (in that the names should indicate what the function does).

Something like:
normalized_symmetric_adjacency_matrix

is what is computed here. It is true (and perhaps) interesting that this is used in Kipf's work, but it is not an invention of that paper as well..
The information where this normalization is used could be provided in the function description with a reference. Examples: Kipf... [REF]

Second, it would be good to have all names consistent, atm we have things like

normalize_x_laplacian (verb first), and ..._adjacency_matrix_normalization.

We should decide for one style and apply consistently, Perhaps something like:

compute_xy_normalized_zmatrix ??

Reply:
I agree.

Reply:
@michaelschaub  do you know a better reference for the above normalization ?

Also, do you have a ref for D^{-1}A normalization? 

please let me know if you like the current naming convection-I can change it again if you have something better in mind.